src/test/util/blocktools.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2019 The Unit-e developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/util/blocktools.h>
+
+namespace blocktools {
+
+CBlockIndex *BlockIndexFake::MakeBlockIndex(const uint256 &hash, CBlockIndex *const prev) {
+  auto result = block_indexes.emplace(hash, CBlockIndex());
+  CBlockIndex *const index = &result.first->second;
+  index->phashBlock = &result.first->first;
+  index->nHeight = prev ? prev->nHeight + 1 : 0;
+  index->pprev = prev;
+  return index;
+}
+
+CBlockIndex *BlockIndexFake::Generate(const std::size_t count, const uint256 &starting_point) {
+  auto starting_block = block_indexes.find(starting_point);
+  std::size_t height = 0;
+  CBlockIndex *const starting_index = [&]() {
+    if (starting_block == block_indexes.end()) {
+      ++height;
+      return MakeBlockIndex(starting_point, nullptr);
+    } else {
+      return &starting_block->second;
+    }
+  }();
+  assert(starting_index);
+  assert(starting_index->phashBlock);
+
+  CBlockIndex *current_index = starting_index;
+  for (; height < count; ++height) {
+    const uint256 hash = GetRandHash();
+    current_index = MakeBlockIndex(hash, current_index);
+    assert(current_index);
+    assert(current_index->phashBlock);
+    assert(current_index->pprev);
+  }
+
+  return current_index;
+}
+
+std::shared_ptr<std::vector<CBlockIndex *>> BlockIndexFake::GetChain(const uint256 &tip_hash) {
```

Comment:
I'd use CBlockIndex instead of the hash.

Reply:
In every usual use case, we don't need `std::shared_ptr`. It's really used only from `SetActiveChain` because it copies shared pointer to closures that increases its lifetime. I suggest return vector here, and make it shared inside `SetActiveChain` only.

Reply:
Changed in edb09171c09b04ab6fe79871e7f8996b79e375af